discovery/azure/azure.go

```diff
@@ -126,60 +128,35 @@ func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
 }
 
 func init() {
-	prometheus.MustRegister(azureSDRefreshDuration)
-	prometheus.MustRegister(azureSDRefreshFailuresCount)
+	prometheus.MustRegister(refreshDuration)
+	prometheus.MustRegister(refreshFailuresCount)
 }
 
-// Discovery periodically performs Azure-SD requests. It implements
-// the Discoverer interface.
 type Discovery struct {
-	cfg      *SDConfig
-	interval time.Duration
-	port     int
-	logger   log.Logger
+	*refresh.Discovery
+	logger log.Logger
+	cfg    *SDConfig
+	port   int
 }
 
 // NewDiscovery returns a new AzureDiscovery which periodically refreshes its targets.
 func NewDiscovery(cfg *SDConfig, logger log.Logger) *Discovery {
 	if logger == nil {
 		logger = log.NewNopLogger()
 	}
-	return &Discovery{
-		cfg:      cfg,
-		interval: time.Duration(cfg.RefreshInterval),
-		port:     cfg.Port,
-		logger:   logger,
-	}
-}
-
-// Run implements the Discoverer interface.
-func (d *Discovery) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
-	ticker := time.NewTicker(d.interval)
-	defer ticker.Stop()
-
-	for {
-		select {
-		case <-ctx.Done():
-			return
-		default:
-		}
-
-		tg, err := d.refresh(ctx)
-		if err != nil {
-			level.Error(d.logger).Log("msg", "Unable to refresh during Azure discovery", "err", err)
-		} else {
-			select {
-			case <-ctx.Done():
-			case ch <- []*targetgroup.Group{tg}:
-			}
-		}
-
-		select {
-		case <-ticker.C:
-		case <-ctx.Done():
-			return
-		}
-	}
+	d := &Discovery{
+		cfg:    cfg,
+		port:   cfg.Port,
+		logger: logger,
+	}
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		time.Duration(cfg.RefreshInterval),
+		d.refresh,
+		refresh.WithDuration(refreshDuration),
```

Comment:
This is a smell, it'd be better to define it in the refresh package and use labels.